src/lib/io/sol-uart-impl-riot.c

```diff
@@ -119,48 +176,107 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
     uart_poweron(uart->id);
     ret = sol_interrupt_scheduler_uart_init_int(uart->id,
         baud_rata_table[config->baud_rate], uart_rx_cb, uart,
-        &uart->async.handler);
+        &uart->handler);
     SOL_INT_CHECK_GOTO(ret, != 0, fail);
 
-    uart->async.rx_cb = config->rx_cb;
-    uart->async.rx_user_data = config->rx_cb_user_data;
-    uart->async.tx_buffer = NULL;
+    uart->rx_cb = config->rx_cb;
+    uart->tx_cb = config->tx_cb;
+    uart->user_data = config->user_data;
+    sol_ptr_vector_init(&uart->pending_blobs);
+
+    rx_size = config->rx_size;
+    if (!rx_size)
+        rx_size = DEFAULT_BUFFER_SIZE;
+    else
+        flags |= SOL_BUFFER_FLAGS_FIXED_CAPACITY;
+
+    uart->tx_size = config->tx_size;
+    rx_buf = malloc(rx_size);
+    SOL_NULL_CHECK_GOTO(rx_buf, err_buf);
+    sol_buffer_init_flags(&uart->rx, rx_buf, rx_size, flags);
+
     return uart;
 
+err_buf:
+    sol_interrupt_scheduler_uart_stop(uart->id, uart->handler);
 fail:
+    uart_poweroff(uart->id);
     free(uart);
     return NULL;
 }
 
 SOL_API void
 sol_uart_close(struct sol_uart *uart)
 {
+    struct sol_blob *blob;
+    uint16_t i;
+
     SOL_NULL_CHECK(uart);
 
-    if (uart->async.tx_writer) {
-        sol_timeout_del(uart->async.tx_writer);
-        uart_tx_dispatch(uart, -1);
+    if (uart->in_use) {
+        uart->delete_me = true;
+        return;
     }
 
-    if (uart->async.handler)
-        sol_interrupt_scheduler_uart_stop(uart->id, uart->async.handler);
-    uart_poweroff(uart->id);
+    uart->in_use = true;
+    SOL_PTR_VECTOR_FOREACH_IDX (&uart->pending_blobs, blob, i) {
+        uart_tx_dispatch(uart, blob, -ECANCELED);
+        sol_blob_unref(blob);
+    }
+
+    if (uart->tx_writer)
+        sol_timeout_del(uart->tx_writer);
+
+    if (uart->rx_reader)
+        sol_timeout_del(uart->rx_reader);
 
+    if (uart->handler)
+        sol_interrupt_scheduler_uart_stop(uart->id, uart->handler);
```

Comment:
likewise, you can stop the interruption for the given UART. See my comment for close() in impl-linux